lib/net.js

```diff
@@ -687,6 +663,49 @@ Socket.prototype._write = function(data, encoding, cb) {
   this._writeGeneric(false, data, encoding, cb);
 };
 
+Socket.prototype._end = function _end(cb) {
+  debug('_end');
+
+  // If still connecting - defer handling 'finish' until 'connect' will happen
+  if (this._connecting) {
+    debug('_end: not yet connected');
+    return this.once('connect', function() {
+      this._end(cb);
+    });
+  }
+
+  if (!this.readable || this._readableState.ended) {
+    debug('_end: not readable or ended');
+    return cb();
+  }
```

Comment:
If the readable side has closed, does this mean the shutdown req has already been issued?